frontend/lib/parsing/parsing-queries.cpp

```diff
@@ -317,6 +317,148 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
   return parse(context, path, emptyParentSymbolPath);
 }
 
+const std::vector<ID>& toplevelModulesInFile(Context* context,
+                                             UniqueString path) {
+  QUERY_BEGIN(toplevelModulesInFile, context, path);
+
+  std::vector<ID> result;
+  const ModuleVec& modules = parseToplevel(context, path);
+  for (const Module* mod : modules) {
+    result.push_back(mod->id());
+  }
+
+  return QUERY_END(result);
+}
+
+struct FindMain {
+  std::vector<const Function*> mainProcsFound;
+  std::vector<const Module*> modulesFound;
+
+  FindMain(Context* context) { }
+
+  bool enter(const Function* fn) {
+    if (fn->name() == USTR("main")) {
+      mainProcsFound.push_back(fn);
+    }
+    return true;
+  }
+  void exit(const Function* fn) { }
+
+  bool enter(const Module* mod) {
+    modulesFound.push_back(mod);
+    return true;
+  }
+  void exit(const Module* mod) { }
+
+  // traverse through anything else
+  bool enter(const AstNode* ast) {
+    return true;
+  }
+  void exit(const AstNode* ast) { }
+};
+
+
+static const ID& findMainModuleImpl(Context* context,
+                                    std::vector<ID> commandLineModules,
+                                    UniqueString requestedMainModuleName) {
+  QUERY_BEGIN(findMainModuleImpl, context,
+              commandLineModules, requestedMainModuleName);
+
+  ID result;
+  auto findMain = FindMain(context);
+
+  // traverse to find modules (to check name) and main functions
+  for (const auto& id : commandLineModules) {
+    if (const AstNode* ast = idToAst(context, id)) {
+      ast->traverse(findMain);
+    }
+  }
+
+  if (!requestedMainModuleName.isEmpty()) {
+    // use requestedMainModuleName to get the main module name
+    const Module* matchingModule = nullptr;
+    for (const Module* mod : findMain.modulesFound) {
+      if (mod->name() == requestedMainModuleName) {
+        matchingModule = mod;
+      }
+    }
+    if (matchingModule) {
+      result = matchingModule->id();
+    } else {
+      context->error(Location(),
+                     "could not find module named '%s' for --main-module",
+                     requestedMainModuleName.c_str());
```

Comment:
should you return here to avoid guessing the main module when the user explicitly specified one?